core/includes/file.inc

```diff
@@ -2018,6 +2021,74 @@ function file_validate_image_orientation(File $file, $orientation = 0) {
   return $errors;
 }
 
+/**
+ * Validate uploaded SVG files.
+ *
+ * @param File $file
+ *   A file entity.
+ *
+ * @return array
+ *   If the file is an SVG and is either not valid or contains
+ *   dangerous content, the array will contain an error message.
+ *
+ * @see hook_file_validate()
+ */
+function file_validate_svg(File $file) {
+  $errors = array();
+
+  if ($file->filemime == 'image/svg+xml') {
+    $file_contents = file_get_contents($file->uri);
+    libxml_use_internal_errors(TRUE);
+    $xml = simplexml_load_string($file_contents);
+    $errors = libxml_get_errors();
+
+    if (!$xml || !empty($errors)) {
+      return array(t('Invalid SVG file.'));
+    }
+    else {
+      $errors = array();
+      $namespaces = $xml->getNamespaces();
+      if (!in_array('http://www.w3.org/2000/svg', $namespaces)) {
+        $errors[] = t('Invalid SVG namespace.');
+      }
+
+      $search_patterns = array(
+        // @see https://developer.mozilla.org/en-US/docs/Web/SVG/Element
+        '//svg:foreignObject',
+        '//svg:script',
+        '//html:iframe',
```

Comment:
The "bad" SVG at https://svg.enshrined.co.uk/ has a few other issues that it seems to prefer to sanitize. Maybe we need some more search patterns.

Reply:
This catches the XML nodes with remote fill absolute URLs
XPath:
```
        '//*[contains(@fill, "url") and contains(@fill,"http")]',
        '//*[contains(@fill, "url") and contains(@fill,"//")]',
        '//*[contains(@fill, "url") and contains(@fill,"ftp")]',

```
XML: 
```
<rect fill="url('http://example.com/benis.svg')"....></rect>
<rect fill="url('https://example.com/benis.svg')"....></rect>
<rect fill="  url(  ' https://example.com/benis.svg '  ) "....></rect>
<rect fill="url('ftp://192.168.2.1/benis.svg')"....></rect>
<rect fill="url('//example.com/benis.svg')"....></rect>
<rect fill="url('/benis.svg')"....></rect> (PASSES)
<rect fill="url('#benis.svg')"....></rect> (PASSES)
```

Reply:
I havent researched *why* absolute URLs are sanitized but posting in case we agree to validate for those

Reply:
> I havent researched why absolute URLs are sanitized

That seems to be something, that can get turned on as additional sanitize step. External URLs aren't per se dangerous - have a look at one example usage in the issue - where external URLs to Wikipedia were set. We don't need that anymore for our field formatter, as we don't display svg inline anymore, though.

I'd suggest to discuss the "reject external links or all links" more in depth on the issue, as that isn't just a technical topic. :wink:

Reply:
And maybe it's not a blocker either. We can tighten up the validation later.